src/httpserver.cpp

```diff
@@ -149,10 +150,58 @@ static GlobalMutex g_httppathhandlers_mutex;
 static std::vector<HTTPPathHandler> pathHandlers GUARDED_BY(g_httppathhandlers_mutex);
 //! Bound listening sockets
 static std::vector<evhttp_bound_socket *> boundSockets;
+
+/**
+ * @brief Helps keep track of open `evhttp_connection`s with active `evhttp_requests`
+ *
+ */
+class HTTPRequestTracker
+{
+private:
+    mutable Mutex m_mutex;
+    mutable std::condition_variable m_cv;
+    //! For each connection, keep a counter of how many requests are open
+    std::unordered_map<const evhttp_connection*, size_t> m_tracker GUARDED_BY(m_mutex);
+
+    void RemoveConnectionInternal(const decltype(m_tracker)::iterator it) EXCLUSIVE_LOCKS_REQUIRED(m_mutex)
+    {
+        m_tracker.erase(it);
+        if (m_tracker.empty()) m_cv.notify_all();
+    }
+public:
+    //! Increase request counter for the associated connection by 1
+    void AddRequest(evhttp_request* req)
```

Comment:
You need this:

```diff
diff --git i/src/httpserver.cpp w/src/httpserver.cpp
index d9203f8715..069511563c 100644
--- i/src/httpserver.cpp
+++ w/src/httpserver.cpp
@@ -167,37 +167,40 @@ private:
     {
         m_tracker.erase(it);
         if (m_tracker.empty()) m_cv.notify_all();
     }
 public:
     //! Increase request counter for the associated connection by 1
-    void AddRequest(evhttp_request* req)
+    void AddRequest(evhttp_request* req) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
     {
         const evhttp_connection* conn{Assert(evhttp_request_get_connection(Assert(req)))};
         WITH_LOCK(m_mutex, ++m_tracker[conn]);
     }
     //! Decrease request counter for the associated connection by 1, remove connection if counter is 0
-    void RemoveRequest(evhttp_request* req)
+    void RemoveRequest(evhttp_request* req) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
     {
         const evhttp_connection* conn{Assert(evhttp_request_get_connection(Assert(req)))};
         LOCK(m_mutex);
         auto it{m_tracker.find(conn)};
         if (it != m_tracker.end() && it->second > 0) {
             if (--(it->second) == 0) RemoveConnectionInternal(it);
         }
     }
     //! Remove a connection entirely
-    void RemoveConnection(const evhttp_connection* conn)
+    void RemoveConnection(const evhttp_connection* conn) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
     {
         LOCK(m_mutex);
         auto it{m_tracker.find(Assert(conn))};
         if (it != m_tracker.end()) RemoveConnectionInternal(it);
     }
-    size_t CountActiveConnections() const { return WITH_LOCK(m_mutex, return m_tracker.size()); }
+    size_t CountActiveConnections() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        return WITH_LOCK(m_mutex, return m_tracker.size());
+    }
     //! Wait until there are no more connections with active requests in the tracker
-    void WaitUntilEmpty() const
+    void WaitUntilEmpty() const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
     {
         WAIT_LOCK(m_mutex, lock);
         m_cv.wait(lock, [this]() EXCLUSIVE_LOCKS_REQUIRED(m_mutex) { return m_tracker.empty(); });
     }
 };
 //! Track active requests
```

`Mutex` has stronger requirements/guarantees. Or, in other words, `GlobalMutex` is a `Mutex` with silenced warnings because of whatever reasons.

Reply:
Thank you, fixed!